src/coap_mbedtls.c

```diff
@@ -0,0 +1,1825 @@
+  }
+  return ret;
+}
+
+#if defined(MBEDTLS_SSL_SRV_C)
+
+static int setup_server_ssl_session(coap_session_t *c_session,
+                                    coap_mbedtls_env_t *m_env)
+{
+  coap_mbedtls_context_t *m_context =
+           (coap_mbedtls_context_t *)c_session->context->dtls_context;
+  int ret = 0;
+  m_context->psk_pki_enabled |= IS_SERVER;
+
+  mbedtls_ssl_cookie_init(&m_env->cookie_ctx);
+  if ((ret = mbedtls_ssl_config_defaults(&m_env->conf,
+                  MBEDTLS_SSL_IS_SERVER,
+                  c_session->proto == COAP_PROTO_DTLS ?
+                   MBEDTLS_SSL_TRANSPORT_DATAGRAM :
+                   MBEDTLS_SSL_TRANSPORT_STREAM,
+                  MBEDTLS_SSL_PRESET_DEFAULT)) != 0) {
+    coap_log(LOG_ERR, "mbedtls_ssl_config_defaults returned -0x%x\n", -ret);
+    goto fail;
+  }
+
+  mbedtls_ssl_conf_rng(&m_env->conf, mbedtls_ctr_drbg_random, &m_env->ctr_drbg);
+
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+  mbedtls_ssl_conf_handshake_timeout(&m_env->conf, 1000, 60000);
+
+  if (m_context->psk_pki_enabled & IS_PSK) {
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+    mbedtls_ssl_conf_psk_cb(&m_env->conf, psk_server_callback, c_session);
+    if (c_session->context->spsk_setup_data.validate_sni_call_back) {
+      mbedtls_ssl_conf_sni(&m_env->conf, psk_sni_callback, c_session);
+    }
+#endif /* MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED */
+  }
+#endif /* MBEDTLS_SSL_PROTO_DTLS */
+
+  if (m_context->psk_pki_enabled & IS_PKI) {
+    ret = setup_pki_credentials(&m_env->cacert, &m_env->public_cert,
+                                &m_env->private_key, m_env, m_context,
+                                c_session, &m_context->setup_data,
+                                COAP_DTLS_ROLE_SERVER);
+    if (ret < 0) {
+      coap_log(LOG_ERR, "PKI setup failed\n");
+      return ret;
+    }
+    if (m_context->setup_data.validate_sni_call_back) {
+      mbedtls_ssl_conf_sni(&m_env->conf, pki_sni_callback, c_session);
+    }
+  }
+
+  if ((ret = mbedtls_ssl_cookie_setup(&m_env->cookie_ctx,
+                                  mbedtls_ctr_drbg_random,
+                                  &m_env->ctr_drbg)) != 0) {
+    coap_log(LOG_ERR, "mbedtls_ssl_cookie_setup: returned -0x%x\n", -ret);
+    goto fail;
+  }
+
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+  mbedtls_ssl_conf_dtls_cookies(&m_env->conf, mbedtls_ssl_cookie_write,
+                                mbedtls_ssl_cookie_check,
+                                &m_env->cookie_ctx );
+#if MBEDTLS_VERSION_NUMBER >= 0x02100100
+  mbedtls_ssl_set_mtu(&m_env->ssl, c_session->mtu);
+#endif /* MBEDTLS_VERSION_NUMBER >= 0x02100100 */
+#endif /* MBEDTLS_SSL_PROTO_DTLS */
+fail:
+  return ret;
+}
+#endif /* MBEDTLS_SSL_SRV_C */
+
+#define MAX_CIPHERS 100
+static int psk_ciphers[MAX_CIPHERS];
+static int pki_ciphers[MAX_CIPHERS];
+static int processed_ciphers = 0;
+
+static void
+set_ciphersuites(mbedtls_ssl_config *conf, int is_psk)
```

Comment:
I think it would be good to have more descriptive enums here to distinguish between PSK and PKI mode.

Reply:
Now using enum